src/hypothesis/extra/datetime.py

```diff
@@ -23,62 +32,185 @@
+                if not self.timezones or self.allow_naive and cu.boolean(data):
+                    return result
+                to_tz = cu.choice(data, self.timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                # Aware bounds: localise to UTC, then convert properly
+                return to_tz.normalize(
+                    pytz.utc.localize(result).astimezone(to_tz))
             except (OverflowError, ValueError):
+                # Overflow in .astimezone(); or invalid month/days combination
                 pass
 
 
 @defines_strategy
-def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None):
+def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None,
+              min_datetime=None, max_datetime=None):
     """Return a strategy for generating datetimes.
 
     allow_naive=True will cause the values to sometimes be naive.
     timezones is the set of permissible timezones. If set to an empty
-    collection all timezones must be naive. If set to None all available
+    collection all timezones will be naive. If set to None all available
     timezones will be used.
 
+    There are two ways to set bounds on the values from this strategy:
+    timezone-aware and naive.
+
+    If two timezone-aware datetime objects are used as the bounds, all
+    values drawn from the strategy will be timezone-aware.  The moment
+    in time represented by a drawn value will be between the moments
+    represented by the bounds.  Because values may be represented in
+    any of the supplied timezones, a naive comparison may show that a value
+    is out of bounds - for example, 10am in London is actually earlier than
+    8am in New York.
+
+    If two naive date or datetime objects are used as the bounds, values
+    drawn from the strategy may or may not be timezone-aware.
+    All values will lie between the bounds when their tzinfo is ignored.
+    If a date object is used, time information will be added to allow the
+    widest possible range of values.  The min_year and max_year arguments
+    are deprecated, but if given they allow the widest range of values
+    possible in the same way as dates.
+
+    For example:
+
+    .. code-block:: pycon
+
+      >>> from hypothesis.extra.datetime import datetimes
+      >>> datetimes().example()
+      datetime.datetime(1705, 1, 20, 0, 32, 0, 973139, tzinfo=<DstTzInfo 'Is...
+      >>> datetimes().example()
+      datetime.datetime(7274, 6, 9, 23, 0, 31, 75498, tzinfo=<DstTzInfo 'Ame...
+
+    As you can see, it produces years from quite a wide range. If you want to
+    narrow it down you can ask for a more specific range of years:
+
+    .. code-block:: pycon
+
+      >>> import datetime
+      >>> datetimes(max_datetime=datetime.date(2000, 1, 1),
+      >>>           max_datetime=datetime.datetime.now()).example()
```

Comment:
...I may try to get doctests passing for this module, to catch this kind of stupid error.  But only after the implementation is nailed down!